File: packages/python/elizaos/features/advanced_capabilities/clipboard/actions/append.py
"""Clipboard append action -- append content to an existing entry."""

from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Action, ActionResult, Content, ModelType
from elizaos.utils.xml import parse_key_value_xml

from ..services.clipboard_service import create_clipboard_service

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

EXTRACT_TEMPLATE = """Extract the clipboard entry ID and content to append from the user's message.

User message: {text}

Available clipboard entries:
{entries}

Respond with XML containing:
- id: The ID of the clipboard entry to append to (required)
- content: The new content to append (required)

<response>
<id>entry-id</id>
<content>Content to append</content>
</response>"""


async def _validate(runtime: IAgentRuntime, message: Memory, _state: State | None = None) -> bool:
    text = (message.content.text or "").lower()
    return "clipboard" in text and "append" in text


async def _handler(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult:
    service = create_clipboard_service(runtime)
    entries = await service.list()
    entries_ctx = "\n".join(f'- {e.id}: "{e.title}"' for e in entries)

    if not entries:
        if callback:
            await callback(
                Content(text="There are no clipboard entries to append to. Create one first.")
            )
        return ActionResult(text="No entries available", success=False)

    prompt = EXTRACT_TEMPLATE.format(text=message.content.text or "", entries=entries_ctx)
    result = await runtime.use_model(ModelType.TEXT_SMALL, prompt=prompt)
    parsed = parse_key_value_xml(str(result))

    if (
        not parsed
        or not isinstance(parsed.get("id"), str)
        or not isinstance(parsed.get("content"), str)
    ):
        if callback:
            await callback(
                Content(
                    text=f"I couldn't determine which note to update or what to add. Available entries:\n{entries_ctx}"
                )
            )
        return ActionResult(text="Failed to extract append info", success=False)

    try:
        entry_id = str(parsed["id"])
        if not await service.exists(entry_id):
            if callback:
                await callback(
                    Content(
                        text=f'Clipboard entry "{entry_id}" not found. Available entries:\n{entries_ctx}'
                    )
                )
            return ActionResult(text="Entry not found", success=False)

        existing = await service.read(entry_id)
        from ..types import ClipboardWriteOptions

        entry = await service.write(
            existing.title,
            str(parsed["content"]),
            ClipboardWriteOptions(append=True, tags=existing.tags),
        )

        msg = f'Successfully appended content to "{entry.title}" ({entry.id}).'
        if callback:
            await callback(Content(text=msg, actions=["CLIPBOARD_APPEND_SUCCESS"]))
        return ActionResult(text=msg, success=True)
    except Exception as e:
        if callback:
            await callback(Content(text=f"Failed to append to the note: {e}"))
        return ActionResult(text="Failed to append to clipboard entry", success=False)


clipboard_append_action = Action(
    name="CLIPBOARD_APPEND",
    similes=["APPEND_NOTE", "ADD_TO_NOTE", "CLIPBOARD_ADD"],
    description="Append content to an existing clipboard entry",
    validate=_validate,
    handler=_handler,
    examples=[],
)

File: packages/typescript/src/features/advanced-capabilities/clipboard/actions/append.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	ModelType,
	parseKeyValueXml,
	type State,
} from "../../../../types/index.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

interface AppendInput {
	id: string;
	content: string;
}

function isValidAppendInput(obj: Record<string, unknown>): boolean {
	return (
		typeof obj.id === "string" &&
		obj.id.length > 0 &&
		typeof obj.content === "string" &&
		obj.content.length > 0
	);
}

const EXTRACT_TEMPLATE = `Extract the clipboard entry ID and content to append from the user's message.

User message: {{text}}

Available clipboard entries:
{{entries}}

Respond with XML containing:
- id: The ID of the clipboard entry to append to (required)
- content: The new content to append (required)

<response>
<id>entry-id</id>
<content>Content to append</content>
</response>`;

async function extractAppendInfo(
	runtime: IAgentRuntime,
	message: Memory,
	availableEntries: string,
): Promise<AppendInput | null> {
	const prompt = EXTRACT_TEMPLATE.replace(
		"{{text}}",
		message.content.text ?? "",
	).replace("{{entries}}", availableEntries);

	const result = await runtime.useModel(ModelType.TEXT_SMALL, {
		prompt,
		stopSequences: [],
	});

	logger.debug("[ClipboardAppend] Extract result:", result);

	const parsed = parseKeyValueXml(String(result)) as Record<
		string,
		unknown
	> | null;

	if (!parsed || !isValidAppendInput(parsed)) {
		logger.error("[ClipboardAppend] Failed to extract valid append info");
		return null;
	}

	return {
		id: String(parsed.id),
		content: String(parsed.content),
	};
}

const spec = requireActionSpec("CLIPBOARD_APPEND");

export const clipboardAppendAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avTextRaw =
			typeof message?.content?.text === "string" ? message.content.text : "";
		const __avText = __avTextRaw.toLowerCase();
		const __avKeywords = ["clipboard", "append"];
		const __avKeywordOk =
			__avKeywords.length > 0 &&
			__avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
		const __avRegex = /\b(?:clipboard|append)\b/i;
		const __avRegexOk = __avRegex.test(__avText);
		const __avSource = String(message?.content?.source ?? "");
		const __avExpectedSource = "";
		const __avSourceOk = __avExpectedSource
			? __avSource === __avExpectedSource
			: Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
		const __avOptions = options && typeof options === "object" ? options : {};
		const __avInputOk =
			__avText.trim().length > 0 ||
			Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
			Boolean(message?.content && typeof message.content === "object");

		if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
			return false;
		}

		const __avLegacyValidate = async (
			_runtime: IAgentRuntime,
			_message: Memory,
		): Promise<boolean> => {
			return true;
		};
		try {
			return Boolean(await __avLegacyValidate(runtime, message));
		} catch {
			return false;
		}
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		const service = createClipboardService(runtime);

		// Get list of available entries for context
		const entries = await service.list();
		const entriesContext = entries
			.map((e) => `- ${e.id}: "${e.title}"`)
			.join("\n");

		if (entries.length === 0) {
			if (callback) {
				await callback({
					text: "There are no clipboard entries to append to. Create one first with CLIPBOARD_WRITE.",
					actions: ["CLIPBOARD_APPEND_EMPTY"],
					source: message.content.source,
				});
			}
			return { success: false, text: "No entries available" };
		}

		const appendInfo = await extractAppendInfo(
			runtime,
			message,
			entriesContext,
		);

		if (!appendInfo) {
			if (callback) {
				await callback({
					text: `I couldn't determine which note to update or what to add. Available entries:\n${entriesContext}`,
					actions: ["CLIPBOARD_APPEND_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to extract append info" };
		}

		try {
			// Check if entry exists
			const exists = await service.exists(appendInfo.id);
			if (!exists) {
				if (callback) {
					await callback({
						text: `Clipboard entry "${appendInfo.id}" not found. Available entries:\n${entriesContext}`,
						actions: ["CLIPBOARD_APPEND_NOT_FOUND"],
						source: message.content.source,
					});
				}
				return { success: false, text: "Entry not found" };
			}

			// Get existing entry to preserve title
			const existingEntry = await service.read(appendInfo.id);

			// Write with append option
			const entry = await service.write(
				existingEntry.title,
				appendInfo.content,
				{
					append: true,
					tags: existingEntry.tags,
				},
			);

			const successMessage = `Successfully appended content to "${entry.title}" (${entry.id}).`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_APPEND_SUCCESS"],
					source: message.content.source,
				});
			}

			return { success: true, text: successMessage, entry };
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardAppend] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to append to the note: ${errorMsg}`,
					actions: ["CLIPBOARD_APPEND_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to append to clipboard entry" };
		}
	},

	examples: [],
};

export default clipboardAppendAction;
